tests/test_env.py

```diff
@@ -212,22 +213,57 @@ def test_skip_gtiff(gdalenv):
 @mingdalversion
 @credentials
 @pytest.mark.network
-def test_s3_open_with_session(gdalenv):
+def test_s3_open_with_env(gdalenv):
     """Read from S3 demonstrating lazy credentials."""
-    with rasterio.open(L8TIF) as dataset:
-        assert dataset.count == 1
+    with rasterio.Env():
+        with rasterio.open(L8TIF) as dataset:
+            assert dataset.count == 1
 
 
 @mingdalversion
 @credentials
 @pytest.mark.network
-def test_s3_open_with_default_session(gdalenv):
+def test_s3_open_with_implicit_env(gdalenv):
     """Read from S3 using default env."""
     with rasterio.open(L8TIF) as dataset:
         assert dataset.count == 1
 
 
 @mingdalversion
+@credentials
+@pytest.mark.network
+def test_env_open_s3(gdalenv):
+    """Read using env as context."""
+    creds = boto3.Session().get_credentials()
+    with rasterio.Env(aws_access_key_id=creds.access_key,
+                      aws_secret_access_key=creds.secret_key):
+        with rasterio.open(L8TIF) as dataset:
+            assert dataset.count == 1
+
+
+@mingdalversion
+@credentials
+@pytest.mark.network
+def test_env_open_s3_credentials(gdalenv):
+    """Read using env as context."""
+    aws_session = boto3.Session()
+    with rasterio.Env(aws_session=aws_session):
+        with rasterio.open(L8TIF) as dataset:
+            assert dataset.count == 1
+
+
+@mingdalversion
+@credentials
+@pytest.mark.network
+def test_ensured_env_no_credentializing(gdalenv):
+    """open's extra env doesn't override outer env"""
+    with rasterio.Env(aws_access_key_id='foo',
+                      aws_secret_access_key='bar'):
+        with pytest.raises(Exception):
+            rasterio.open(L8TIFB2)
```

Comment:
This test was failing, meaning the URL was being opened even with bad credentials, and then I realized it was because of VSI caching. By changing the URL, we get the failure we expect.